### PR TITLE
LOG-1772: Canoe: productionise Paddle

### PR DESCRIPTION
**Pull request by pedrocunha**
Still work in progress, but just opening the PR to create some basic comments

===

Jira story [#LOG-1772](https://deliveroo.atlassian.net/browse/LOG-1772) in project *Logistics*:

> We have created https://github.com/deliveroo/paddle as a proof of concept.
> 
> Start master from scratch, and issue a PR with the current functionality and proper tests.
> 
> The expected functionality is:
> 
> - a command to store a folder to S3 according to the versioning structure described in https://docs.google.com/document/d/1vtDMf50ak170_isB7-761ZsJZroGnbH9j6vd10RkXiY
> - a command to retrieve a folder from S3 given a model name, version and optional branch

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,40 @@
...
+[default]
+aws_access_key_id=xxx
+aws_secret_access_key=yyy
+region=eu-west-1
+```
+
+
+```
+$ go build
+```
```
Comment (humzashah) [inline README.md]: we should specify that Go must be installed only for development / tests. for actual usage, we'll be generating a binary which we can include instructions in the Readme for.

File: cli/data/get.go [hunk 2]
```diff
@@ -27,6 +27,21 @@ import (
 	"strings"
 )
 
+type S3Target struct {
+	bucket string
+	prefix string
+	path string
+}
+
+func (s *S3Target) copy() *S3Target {
```
Comment (humzashah) [inline cli/data/get.go]: is this here to ensure that we aren't mutating?

Comment (pedrocunha) [reply cli/data/get.go]: I don't understand enough yet about Go to grasp if I should be just passing structs by value and allow the runtime to clone those from my understanding but yet.. here was just do that scenario where you want to be a good citzen:

```ruby
class A
  def initialize(options = {})
     opts = options.dup
     opts.merge(...)
  end
end
```

File: cli/data/get.go [hunk 3]
```diff
@@ -53,69 +75,80 @@ func init() {
...
-		path = fmt.Sprintf("%s/%s/%s", version, branch, path)
+	/*
+	 * HEAD contains the path to latest folder
+	 */
+	if source.path == "HEAD" {
+		source = source.copy()
+		source.path = readHEAD(session, source)
+	}
+
+	fmt.Println("Copying " + source.fullPath() + " to " + destination)
```
Comment (me) [inline cli/data/get.go]: I kinda liked a more minimal output to be honest, no biggie though

File: cli/data/get.go (hunk 2 above)
Comment (me) [reply cli/data/get.go]: Yeah I wouldn't have this function, if anything we can just copy below; and `copy` is a built-in so it's not a great name. I would just pass by value, no reason to pass by reference here.

File: README.md (hunk 1 above)
Comment (pedrocunha) [reply README.md]: 👍

File: README.md [hunk 4]
```diff
@@ -1,8 +1,40 @@
...
+You will also need to create a `$HOME/.aws/config` or `$HOME/.aws/credentials` so Paddle can connect to AWS, e.g.:
+
+```
+> cat $HOME/.aws/credentials
+[default]
+aws_access_key_id=xxx
+aws_secret_access_key=yyy
+region=eu-west-1
+```
+
```
Comment (jeffreylo) [inline README.md]: unnecessary newline

File: README.md [hunk 5]
```diff
@@ -1,8 +1,40 @@
...
+cd ~/go/src/github.com/deliveroo
+git clone git@github.com:deliveroo/paddle.git
+cd paddle
+```
+
+You will need create a `$HOME/.paddle.yaml` that contains the bucket name, e.g:
+
+```
+> cat $HOME/.paddle.yaml
+bucket: roo-bucket
```
Comment (me) [inline README.md]: Might be worth pointing out that these settings can be configured through ENV as well (e.g. BUCKET).

File: README.md [hunk 6]
```diff
@@ -21,6 +21,8 @@ You will need create a `$HOME/.paddle.yaml` that contains the bucket name, e.g:
 bucket: roo-bucket
 ```
 
+or if you prefer specific `BUCKET` as an environment variable
```
Comment (me) [inline README.md]: Typo: "specify"

**Pull request merged**